lib/functional/func_collections.nit

```diff
@@ -0,0 +1,106 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Copyright 2019 Louis-Vincent Boudreault <lv.boudreault95@gmail.com>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# In place data manipulation inside an hashmap
+#
+# Provides a Monad-like structure to do computation over hashmap entries.
+# Warning: The Entry API is not thread-safe, i.e concurrent entry manipulation
+# might put the Map in a inconsistent state. In multithreading environment, you
+# should use guards before creating an `Entry`.
```

Comment:
Is this the same kind of API than Rust's [`Result`](https://doc.rust-lang.org/std/result/)? If so you should link to it and/or still their documentation :p